apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1776,82 +1751,63 @@ func (u *UniterAPI) WatchRelationUnits(ctx context.Context, args params.Relation
-		return unit, nil
+		err = u.oneSetRelationStatus(ctx, unitName, arg.RelationId,
+			arg.Status, arg.Message)
+		results[i].Error = apiservererrors.ServerError(err)
 	}
+	statusResults.Results = results
+	return statusResults, nil
+}
 
-	checker := u.leadershipChecker
-	changeOne := func(arg params.RelationStatusArg) error {
-		// TODO(wallyworld) - the token should be passed to SetStatus() but the
-		// interface method doesn't allow for that yet.
-		unitTag := arg.UnitTag
-		if unitTag == "" {
-			// Older clients don't pass in the unit tag explicitly.
-			unitTag = u.auth.GetAuthTag().String()
-		}
-		unit, err := getUnit(unitTag)
-		if err != nil {
-			return err
-		}
-		token := checker.LeadershipCheck(unit.ApplicationName(), unit.Name())
-		if err := token.Check(); err != nil {
-			return errors.Trace(err)
-		}
+func (u *UniterAPI) oneSetRelationStatus(
+	ctx context.Context,
+	unitName coreunit.Name,
+	relID int,
+	relStatus params.RelationStatusValue,
+	message string,
+) error {
+	// Verify the relation exist before continuing.
+	relationUUID, err := u.relationService.GetRelationUUIDByID(ctx, relID)
+	if errors.Is(err, errors.NotFound) {
+		return apiservererrors.ErrPerm
+	} else if err != nil {
+		return internalerrors.Capture(err)
+	}
 
-		rel, err := u.st.Relation(arg.RelationId)
-		if errors.Is(err, errors.NotFound) {
-			return apiservererrors.ErrPerm
-		} else if err != nil {
-			return errors.Trace(err)
-		}
-		_, err = rel.Unit(unit)
-		if errors.Is(err, errors.NotFound) {
-			return apiservererrors.ErrPerm
-		} else if err != nil {
-			return errors.Trace(err)
+	// If we are transitioning from "suspending" to "suspended",
+	// we retain any existing message so that if the user has
+	// previously specified a reason for suspending, it is retained.
+	if message == "" && relStatus == params.Suspended {
+		current, err := u.relationService.GetRelationStatus(ctx, relationUUID)
+		if err != nil {
+			return internalerrors.Capture(err)
 		}
-		// If we are transitioning from "suspending" to "suspended",
-		// we retain any existing message so that if the user has
-		// previously specified a reason for suspending, it is retained.
-		message := arg.Message
-		if message == "" && arg.Status == params.Suspended {
-			current, err := rel.Status()
-			if err != nil {
-				return errors.Trace(err)
-			}
-			if current.Status == status.Suspending {
-				message = current.Message
-			}
+		if current.Status == status.Suspending {
+			message = current.Message
 		}
-		return rel.SetStatus(status.StatusInfo{
-			Status:  status.Status(arg.Status),
-			Message: message,
-		})
 	}
-	results := make([]params.ErrorResult, len(args.Args))
-	for i, arg := range args.Args {
-		err := changeOne(arg)
-		results[i].Error = apiservererrors.ServerError(err)
+	err = u.relationService.SetRelationStatus(ctx, unitName, relationUUID, status.StatusInfo{
+		Status:  status.Status(relStatus),
+		Message: message,
+	})
+	if errors.Is(err, errors.NotFound) {
```

Comment:
Wrong error type.

Reply:
We do not have relation domain error types at this time. The facade changes are being made first. We can be more specific at that time. However differentiating between NotFounds isn't interesting in the facade.

The current facade implementation says that any NotFound returns an ErrPerm. I haven't changed that here.